### ISSUE TITLE
[no-unsafe-return] does not allow return any[] when function is typed as `any`

### ISSUE THREAD

**Issue by HolgerJeromin** [typescript-eslint/typescript-eslint#3549]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unsafe-return": "error"
  }
}
```

```TS
function foo(): any {
    if (Math.random()) {
        let res: any = {};
        return res;
    }
    let resArr: any[] = [];
    resArr.push(true);
    return resArr; // <-- eslint unhappy
}
```

**Expected Result**

`any` is any and the rule should be happy with anything :-)

**Actual Result**

error @typescript-eslint/no-unsafe-return : Unsafe return of an `any[]` typed value.

**Additional Info**

Follow up to #3482 

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.27.0` |
| `@typescript-eslint/parser`        | `4.27.0` |
| `TypeScript`                       | `4.1` |
| `ESLint`                           | `7.27.0` |

**Comment by bradzacher** [comment]
Why are you working with `any` so much?

**Comment by HolgerJeromin** [comment]
> Why are you working with any so much?

So much? We building a framework which is handling various customer data here. 

> You should really use generics, union types or function overloads to correctly type your functions in a type-safe manner.

Moving to generics in our API would collide with the [common mistakes of DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes)
>> `getMeAT<T>(): T`: If a type parameter does not appear in the types of any parameters, you don't really have a generic function, you just have a disguised type assertion. Prefer to use a real type assertion, e.g. `getMeAT() as number`.

Internally we could move to `unknown`. 
But breaking our public API (with `any` at some places like `export function resolveDefault(schema: JsonSchema): any;`) without a good reason is nothing we will do.
So forcing our customer from the current `let myNum: number = getMeAT()` to  `let myNum = getMeAT() as number` is no option.
